### ISSUE TITLE
Dapper v2 planning and discussion

### ISSUE THREAD

**Issue by NickCraver** [StackExchange/dapper-dot-net#688]
It's time for a major release and we've just been busy with other libraries. In context though, Dapper is far less work to get there. `netstandard` is already supported and we don't have a long list of desired (breaking) changes to the APIs. Starting a list here to track:

### API changes **(breaking)**:
- [ ] Optional `CancellationToken` params on all Async methods (related: #484)
- [ ] Dapper.Contrib `Insert<T>` and `Insert<T, TKey>` changes for `ISqlAdapter` (discussion in #640, related: #559, #587)
- [ ] Dapper.Contrib - Allow custom `ISqlAdapter` (#497) (note: this locks us down for any future interface additions)
- [ ] `SqlBuilder` needs a *lot* of tests and love (#573, #647)
- [ ] Overload reduction via `CommandFlags` (related: #474)

### Package Changes **(breaking)**:
- [ ] Re-title to "Dapper" & rename Repo (#683)
- [ ] Strong-Name for `Dapper` (deprecating [`Dapper.StrongName`](https://www.nuget.org/packages/Dapper.StrongName/))
- [ ] Strong-Name for `Dapper.Contrib` (#430, #685)

### Missing APIs (non-breaking):
- [ ] `QueryAsync<T>` and related have a few missing from the single/first additions
- [ ] Dapper.Rainbox `netstandard` support (#676)

### PRs to examine:
- [ ] #471 - Allow custom TypeHandlers for standard types (and `.SetValue()`) (related: #206, #259, #303, #319, #433, #571, #607)
- [ ] #458 - Custom type handlers for enums are ignored
- [ ] #512 - Fixed #497 by exposing SqlMapperExtensions.AdapterDictionary as public
- [ ] #558 - Temporary fix (really want to fix the actual issue: #555)
- [ ] #597 - Allow multi-map async queries to be unbuffered (related: #596)
- [ ] #636 - Dapper.Contrib - Support for composite keys in `Get` method

### Cleanup
- [ ] Moving to VS 2017 RC and new `.csproj` system (off Preview 2 `project.json` tooling)
- [ ] Organize folders (after PRs!): `src`, `tests`, `docs`, etc. (and the same in the solution)
- [ ] Release notes/documentation: moving into `/docs` ([GitHub feature post here](https://github.com/blog/2228-simpler-github-pages-publishing), note: this allows PRs to update docs easily too)
- [ ] Global: Interpolated String usage (#651)
- [ ] Firebird Async support (#651 addition)
- [ ] #621 - Static property issues on params
- [ ] #613 - SimpleMemberMap became internal between 1.42 and 1.50.2
- [ ] #551 - Add example for literal replacement to Readme.md
- [ ] #480 - Documentation for SqlBuilder
- [ ] #451 - Clarify memory documentation
- [ ] #417 - Audit where `DbCommand.Cancel()` is called

### Discussion - Should we add a `[Column]` attribute?
Related issues: #447, #463, #515, #605, #623, #670, #681, #572
Potentially covered: #482, #571
There's a lot of demand for naming your columns something different in the database than the POCOs. Message received. So how do we fix this in a clean way? In my view, this cannot be a Dapper.Contrib only fix, since *de*serialization also needs to support the same names. This allows very specific name mappings, backwards compatibility ease (as you rename DB columns, etc.), and less magic like underscore mapping options.

But we also can't rely on an attribute *only*. The attribute must populate a general mapping which is itself editable. This is because in many cases, you cannot touch the POCO to add such an attribute. I think this would have to live in Dapper core and be supported by .Contrib in the `.Insert()`, etc. methods. Thoughts?

### Discussion - Should we add a `DynamicParameters.Add()` overload in for DB params?
See #659 for details, specifically:
```C#
DynamicParameters.Add(string name, DbType type, ...)
```

...okay that took a while, and it's a lot of info but I'm trying to relate all the issues I see with a common solution path. What did I forget? What's wrong? Thoughts on discussion items above (or pretty much anything of course)?

There will also be some background work I'll take on here. Like being able to more readily test Dapper & Dapper.Contrib against all providers. Luckily Docker has matured enough that having a rich test bed against most providers is finally a reality. I'm working with others in the container and cloud space to make this an option for all, but that's not a blocker by any means.

I haven't even shown this to @mgravell yet. So hopefully he doesn't kinda sorta give me a mild case of death at the next meetup. Due to how # links work...this is the best way to share it for review as well, so here it is.

**Comment by claycephas** [comment]
I don't think https://github.com/StackExchange/Dapper/commit/8aa10a0237f0b419fc063eff95f6f01b617cb4c3 alone is a breaking change since it is additive. And from the testing I've done, it is a backward-compatible solution to #471.

Would it be possible to get a new version pushed to NuGet that is simply 1.50.2 + https://github.com/StackExchange/Dapper/commit/8aa10a0237f0b419fc063eff95f6f01b617cb4c3 without including everything else from `master` that may cause breaking changes or waiting on v2?

**Comment by aluanhaddad** [comment]
Personally, I would expect to see all of the public APIs given much improved source level documentation. Specifically, all parameters should have full doc comments, consistent names, and usage examples. I frequently have to read the source code to verify that certain methods behave the way I think they might. 

Even currently existing comments leave much to be desired. For example, `splitOn` has an awful doc comment that reads 
```C#      
/// <param name="splitOn">
/// The Field we should split and read the second object from (default: id)
/// </param>
```

Another example is the `DynamicParameters` constructor that takes an `object` parameter named `template`. This makes one uneasy about assuming it behaves the same way as `void AddDynamicParams(object param)` because... well... why is the name different?
I do not like to mutate objects after initialization and I thought passing a parameter object to the constructor might behave the same way ass `AddDynamicParams`, but I had to read the source code to confirm this because there is not a clear naming pattern and so many methods already have arguments that are weakly typed and optional.

**Comment by paulmrozowski** [comment]
I would LOVE to have the ability to easily access the returned results metadata - things like the original field's width for strings, or scale/precision for numeric types, etc.

**Comment by unsafePtr** [comment]
Hi, as i couldn't open new pull request, so i will leave it here. I think that will be very good to add support of `ValueTuple<>` with named arguments:
```c#
public async Task<(int tickeTypeId, int ticketsSold, decimal price)> Test()
{
    return await context.QueryAsync<(int tickeTypeId, int ticketsSold, decimal price)>(SqlResources.GetCardSales);
}
```
I understood that there is some magic with ValueTuples run-time, but maybe you have an idea how to get it work.

**Comment by RobThree** [comment]
Maybe add some form of query-interception? If we want to do it now we have to implement a Dapper "wrapper" exposing all extension-methods Dapper provides and implement logging there before/after invoking the actual Dapper method.

Maybe a simple event(s) like `BeforeExecute` (and `AfterExecute`) can be raised with the actual `CommandDefinition` as (one of) the events `BeforeExecuteEventArgs` (and `AfterExecuteEventArgs`) argument properties.

So what I'm saying:

```c#
public class BeforeExecuteEventArgs : EventArgs {
    public CommandDefinition CommandDefinition { get; set; }
    // ... maybe more? Maybe context/IDbConnection etc.?
}
```

And then raise the event(s) in the Execute/Query/etc. extensionmethods.

This way logging of queries, for example, can be simplified/more easily be hooked.

**Comment by ricsmania** [comment]
About enum handlers, it would be extremely productive for us if there were a way to add a global handler. For example, most of our enums are declared like this:

    enum DocumentStatus
    {
        Normal = 'N',
        Canceled = 'C'        
    }

And the corresponding tables are char(1) fields with the char value.  The most productive way would be adding a global handler that converts the char value from the enum to the char value from the database.

**Comment by NickCraver** [comment]
@ricsmania Can you explain that one a little bit? The enum declaration you have is equivalent to:
```c#
enum DocumentStatus
{
	Normal = 78,
	Canceled = 67
}
```
...would you expect an implicit `char` translation understanding to integer positions for the enum? Inherently, that has some problems as it could easily be mis-used to trigger enum values that aren't there. For example `(DocumentStatus)320` is perfectly legal, yet practically invalid.

**Comment by ricsmania** [comment]
@NickCraver Right now what we do is we actually use a database function to explicitly convert chars to the ascii value and vice versa. But we don't want to do that conversion using database functions anymore.

What I'm suggesting is actually creating a way to add a global enum handler, so that we could create a single handler that does that converts any enum using whatever rule we define. I'm not suggesting that Dapper should convert the values, but actually that it could provide the means for us to implement our own conversion. Right now it only converts from numeric types it seems.

Also, the current version doesn't work for enum handlers, but testing the suggested solutions it seems we would need to add a handler for each specific enum, which isn't very productive.

**Comment by AndyCJ** [comment]
Hi,

I had a look through the roadmap issues and couldn't see anything related to this, so sorry if I missed it.

The async support is based on the DbConnection class, rather than IDbConnection. And that makes sense since IDbConnection doesn't declare the async methods. This is quite the fail on Microsofts part, but we still have to work with what we have.

The problem that some of us are finding is that there are other connection classes which also expose the async methods, but don't inherit from DbConnection, they just implement IDbConnection.

ReliableSqlConnection is a good example of this. It's pretty much a requirement for working with SqlServer on Azure, or in any other environment where you can encounter transient connection issues.

This is described and discussed in this issue: https://github.com/StackExchange/Dapper/issues/574

Explicitly supporting ReliableSqlConnection would fix the problem, but in a very limited way. Having to explicitly support each new connection class with async methods would be a pain for maintenance and create lots of work for comparatively little gain. I assume a very small fraction of Dapper users desire this.

One of the suggestions in the issue is that Dapper provides an interface which can be implemented to allow using Dapper with custom implementations.

e.g. ReliableDapperConnection : IDapperConnection

Where ReliableDapperConnection  is simply a pass-through class to an instance of ReliableSqlConnection which it contains.

That would allow devs who consume Dapper to get async functionality with just a bit of effort on their part.

**Comment by antgel** [comment]
Hi Dapper team, it would be great to know what the forecast date is for getting v2 out of the door, if there is a forecast. There are things there that look really interesting.

**Comment by CShepartd** [comment]
>it would be great to know what the forecast date is for getting v2 out of the door
Yeah, mby a roadmap?

**Comment by CShepartd** [comment]
.Net Core and Asp.Net Core from 3.0 will support C# 8.0 with nullable reference types, Dapper from 2.0 also should support it

**Comment by MichaelTontchev** [comment]
Can we expect Dapper 2 at any point soon...er or later?

**Comment by NickCraver** [comment]
The current plan is to start later this year.

The major items currently on our plate (read: fighting for time) are:
- Dapper: Needs a lot of project management and code love (v2 release)
- MiniProfiler: At v4, now live and running smooth
- StackExchange.Redis: At v2, but we're fighting some performance and connection issues - these are affecting users unexpectedly and unpredictably in production (so: higher urgency).
- StackExchange.Exceptional: At v2, stable and running smooth
- protobuf-net: No work ATM I'm aware of (C# generators would change this...and Dapper for that matter)
- Porting Stack Overflow to ASP.NET Core (and everything we actually do as part of our jobs)

Overall, my expected order here is:
1. Wrap up StackExchange.Redis work
2. Get Stack Overflow to .NET Core
3. Spin up Dapper v2

The goal is later this year (at least 2H2019) for that to happen, but the list of things to do and sub-tasks is in the hundreds, so it is admittedly a best guess. I'm happy to provide updates at any time though.

**Comment by jnm2** [comment]
Would you be open to breaking the IMemberMap interface or introducing a new more powerful abstract MemberMap class? I'm looking at fixing https://github.com/StackExchange/Dapper/issues/982 (C# tuples encode the eighth item of a tuple as `valueTuple.Rest.Item1` and the 20th item as `valueTuple.Rest.Rest.Item6`). There is no way to support this via IMemberMap. If a more powerful API is not introduced, ValueTuple will have to be special-cased in SqlMapper's IL emit logic.

**Comment by challen0** [comment]
@NickCraver Are you able to provide any updates on the work for Dapper v2? It looks like something I'm having issue with is on the roadmap to be addressed, so I'm just curious if any progress has been made for v2.

**Comment by grahambunce** [comment]
@NickCraver For v3 is there a piece of work to improve extensibility of Dapper.Contrib? e.g. There is no way for me to add/enhance the UpdateAsync<T> method to enable me to add optimistic concurrency; even as simple as adding WHERE Version = @version. The only I've been able to do this easily is to copy/paste the entire SqlMapperExtensions and SqlMapperExtensions.Async classes into my own code and add a new UpdateOptimistic<T> method.

This is because the existing UpdateAsync<T> uses Property caches that are private objects within SqlMapperExtensions so are not available to people trying to add additional functionality for a small area.

Of course the additional of an Optimistic concurrent Update method would solve my specific problem too!

**Comment by TroySchmidt** [comment]
@NickCraver @grahambunce I have been doing some work in a fork of Dommel to add a SqlQuery class.  It basically allows for writing queries with predicates instead of raw SQL.  This adds the benefit to leverage the mappings of columns and table names so that those naming aliases can be configured once in the mappings and then leveraged in the predicates with strong names (properties).  I was thinking about contributing this back to Dapper.Contrib instead, but it seems woefully underpowered as it currently stands.  It seems like for years it has been referenced that Dapper.Contrib functionality is going to be completely overhauled.  

The goal is that this is used internally and then also provides a way to use fluentmap style API to write for custom relationship joins with predicates.

For instance the code below is what I am writing that can then be completely translated with custom table and column names to working SQL.
```
 var sql = _sqlQuery
  .InnerJoin<Child>((p, c) => p.ChildId == c.Id)
  .Select<Parent>(p => new { p.Id, p.Name })
  .Select<Child>(c => new { c.Id, c.Name, c.TypeId})
  .Where(x => x.Name == "Test")
  .Where<Child>(x => x.TypeId == 5)
  .ToSql(out var parameters);
```

**Comment by NickCraver** [comment]
@TroySchmidt I'd say honestly: that's getting to "a different library" territory since it's not our goal to generate complex SQL in Dapper. It's both _very_ complicated to maintain (across DB providers), but also hard to make generic (e.g. arrays in Postgres vs. IN list in SQL vs. string split in newer SQL, etc.). It's also crazy complex when you get into nested children, applies, etc. Abstracting this away a great deal then gets both hard to implement and confusing to use past the simplest case.

Overall, it sounds like something that'd be more aimed at the https://github.com/henkmollema/Dapper-FluentMap style - would suggest proposing an issue there and seeing if they're interested in such? Otherwise: you're getting into Entity Framework territory.

We intend to overhaul the workings of Dapper.Contrib, but not largely the functionality past attribute and code mapping support (e.g. in this issue). The surface area shouldn't be unfamiliar at the end, if that make sense. It should still be low-allocation, etc.

**Comment by petterek** [comment]
Is supporting parameter classes with fields instead of properties something that could/should be considered? 
Just spent 3 hours debugging an insert statement and it all boiled down to my param object used public fields instead of properties..

**Comment by jnm2** [comment]
A Dapper.Analyzers project could solve the usability side of that.

**Comment by TroySchmidt** [comment]
@NickCraver My purpose is to leverage Dapper (and Dommel) to translate GraphQL into SQL.  Dommel is much more flexible with relationships, complex fields, and fluent mapping instead of attributes on the entities themselves like Dapper.Contrib.  It would be nice to utilize only Dapper with a Dapper based GraphQL library that could provide the extensions necessary.   
So is there a way to work on those type of use cases and make sure that Dapper and Dapper.Contrib have the proper improvements to support extensibility for that through other libraries like what I am interested in doing for GraphQL support?

**Comment by mgravell** [comment]
I think that is impossible to comment on without more specifics.

**Comment by TroySchmidt** [comment]
Okay for instance how to implement composite keys in `Dapper.Contrib`?  Then SqlBuilder which could be very powerful, but someone might want to extend the functionality of it with additional libraries cannot do so because it doesn't use interface programming.  So there is no way to override the functionality with extended functionality for things like using the SqlBuilder nomenclature with the functionality of custom property names that Dapper.Contrib or a similar library could add.
Dapper.Contrib does finally allow for custom table name mapping, but not a similar attribute for custom column name mapping to properties.  There is also people that follow separation of concerns and prefer Fluent mapping style implementation and with the attribute method that doesn't provide a way to support both.

**Comment by vitalybibikov** [comment]
@NickCraver 
It would be nice to see an ability to hydrate types with private default ctors and complex custom objects. As there are lots and lots of projects that incorporated DDD style.

Imagine a simple object: Menu
It is quite easy to solve this issue when there is only 1 type in the query.

When you have 1 to many relationship or 1 to many + custom complex object inside (1 to 1 ), it is quite impossible to use and handle all the parameters in ctor.

![image](https://user-images.githubusercontent.com/7008739/81057544-4da41a80-8ed5-11ea-8a2c-5eba47a3c94a.png)

Currently, you either need to decouple your database objects from your domain models or to write impossibly hard-to-read-and-maintain queries.

**Comment by giammin** [comment]
this one:
https://github.com/StackExchange/Dapper/issues/1055

**Comment by NightWatchman** [comment]
Just a quick note about nullable reference types with the current iteration of Dapper (v2.0.78).

Using Dapper in a project, and enabling nullable reference types in that project results in false negatives for the null value checking mechanism as implemented by Microsoft.

For example the method `QuerySingleorDefaultAsync<T>` has the return type `Task<T>`. In this case Microsoft inaccurately assumes that T is never null. This causes the nullable reference type checking mechanism to break down so it cannot be relied upon.

This is a known limitation of the nullable reference type checking when not every single library used has it enabled and configured properly. Doing so in Dapper v3 will obviously fix this.

**Comment by jnm2** [comment]
@NightWatchman It's a little more subtle than that. C# tracks it as neither "never null" nor "maybe null" but a third state, "null-oblivious," due to the method not being declared in a context that has nullability annotations enabled. The effect is the same as what you're saying, in that you won't be warned if you dereference.

**Comment by NightWatchman** [comment]
@jnm2 That makes sense

**Comment by johandanforth** [comment]
Dapper.Contrib support for inserting anonymous objects? Got this suggestion from a fellow dev at work today, Not sure the best way to infer the tablename though, as there obviosly is no TableAttribute. Perhaps as an overload with a table name parameter:

`var id = connection.Insert(new { Name = "Anonymous", Age = 2 }, "Users");`
